File: backend/app/schemas/portal.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict


class PortalBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class DashboardStat(BaseModel):
    label: str
    value: str
    tone: str


class DashboardSummary(BaseModel):
    user_name: str
    headline: str
    learning_streak: int
    stats: List[DashboardStat]
    next_lesson: Optional[Dict[str, Any]]
    recommended_exercise: Optional[Dict[str, Any]]
    active_projects: List[Dict[str, Any]]
    recommendation_panel: List[Dict[str, str]]
    adaptive_focus: Optional[Dict[str, Any]] = None


class DashboardToday(BaseModel):
    focus: List[str]
    highlights: List[str]
    blockers: List[str]


class LessonOut(PortalBase):
    id: int
    learning_path_id: int
    title: str
    slug: str
    summary: str
    content_md: str
    prerequisites_json: List[str]
    estimated_minutes: int
    order_index: int
    tags_json: List[str]
    is_completed: bool = False


class LearningPathOut(PortalBase):
    id: int
    title: str
    slug: str
    description: str
    level: str
    estimated_hours: int
    order_index: int
    is_active: bool
    tags_json: List[str]
    completion_pct: float = 0
    lessons: List[LessonOut] = []


class LessonCompletionResponse(BaseModel):
    lesson_id: int
    completed: bool
    learning_completion_pct: float


class CourseOut(PortalBase):
    id: int
    title: str
    slug: str
    description: str
    difficulty: str
    estimated_hours: int
    milestones_json: List[Dict[str, Any]]
    status: str
    track_focus: str


class CourseProgressIn(BaseModel):
    status: str


class ExerciseOut(PortalBase):
    id: int
    title: str
    slug: str
    category: str
    difficulty: str
    prompt_md: str
    starter_code: str
    solution_code: str
    explanation_md: str
    tags_json: List[str]
    progression_label: Optional[str] = None
    practice_stage: Optional[str] = None
    hint_md: Optional[str] = None
    review_checklist_json: List[str] = []
    success_criteria_json: List[str] = []
    related_lesson_slugs: List[str] = []
    related_lesson_titles: List[str] = []
    next_exercise_id: Optional[int] = None
    next_exercise_slug: Optional[str] = None
    next_exercise_title: Optional[str] = None


class ExerciseAttemptIn(BaseModel):
    submitted_code: str
    notes: str = ""


class ExerciseAttemptOut(PortalBase):
    id: int
    exercise_id: int
    submitted_code: str
    notes: str
    status: str
    score: float
    attempted_at: datetime


class ExerciseDetail(BaseModel):
    exercise: ExerciseOut
    attempts: List[ExerciseAttemptOut]
    review_prompt: Optional[str] = None


class KnowledgeArticleOut(PortalBase):
    id: int
    title: str
    slug: str
    category: str
    summary: str
    content_md: str
    source_links_json: List[str]
    tags_json: List[str]
    last_reviewed_at: datetime


class NewsItemOut(PortalBase):
    id: int
    source_name: str
    title: str
    slug: str
    summary: str
    source_url: str
    category: str
    published_at: datetime
    signal_score: int
    tags_json: List[str]
    is_saved: bool
    is_seeded: bool
    last_synced_at: datetime
    why_it_matters: str
    suggested_action: str
    focus_area: str
    recommended_path_slug: Optional[str] = None
    recommended_path_title: Optional[str] = None
    recommended_exercise_category: Optional[str] = None


class ProjectOut(PortalBase):
    id: int
    title: str
    slug: str
    summary: str
    status: str
    category: str
    stack_json: List[str]
    architecture_md: str
    repo_url: Optional[str]
    demo_url: Optional[str]
    lessons_learned_md: str
    portfolio_score: int


class JobPostingOut(PortalBase):
    id: int
    source_name: str
    title: str
    slug: str
    company_name: str
    location: str
    employment_type: str
    summary: str
    source_url: str
    description_md: str
    published_at: datetime
    tags_json: List[str]
    skill_gaps_json: List[str]
    fit_score: int
    is_saved: bool
    is_seeded: bool
    last_synced_at: datetime
    fit_summary: str
    suggested_action: str
    primary_gap: Optional[str] = None
    focus_area: str
    recommended_path_slug: Optional[str] = None
    recommended_path_title: Optional[str] = None
    recommended_exercise_category: Optional[str] = None


class ProjectIn(BaseModel):
    title: str
    summary: str
    status: str
    category: str
    stack_json: List[str]
    architecture_md: str
    repo_url: Optional[str] = None
    demo_url: Optional[str] = None
    lessons_learned_md: str = ""
    portfolio_score: int = 0


class InterviewQuestionOut(PortalBase):
    id: int
    category: str
    role_type: str
    difficulty: str
    question_text: str
    answer_outline_md: str
    tags_json: List[str]
    practice_count: int = 0
    last_practiced_at: Optional[datetime] = None
    average_confidence: Optional[float] = None


class InterviewPracticeIn(BaseModel):
    confidence_score: int = 3
    notes: str = ""


class InterviewPracticeOut(BaseModel):
    question_id: int
    practice_count: int
    last_practiced_at: datetime
    average_confidence: float


class InterviewRoadmap(BaseModel):
    focus_areas: List[str]
    weekly_plan: List[str]
    rationale: List[str] = []


class PortfolioReadiness(BaseModel):
    overall_score: int
    strongest_signals: List[str]
    gaps_to_close: List[str]
    next_best_moves: List[str]


class SkillGapInsight(BaseModel):
    title: str
    urgency: str
    evidence: str
    action_path: str


class MasteryAreaOut(BaseModel):
    area_slug: str
    area_title: str
    mastery_score: int
    status: str
    evidence: str
    gap: str
    weakest_signal: str
    next_action_path: str
    next_action_label: str


class AdaptiveFocusOut(BaseModel):
    title: str
    reason: str
    action_path: str
    target_kind: str
    mastery_score: int
    area_slug: str
    area_title: str
    action_label: str


class RecommendationOut(BaseModel):
    title: str
    reason: str
    action_path: str
    source_kind: str = "internal"


class JobFitAnalysisOut(BaseModel):
    job_id: int
    fit_score: int
    strengths: List[str]
    gaps: List[str]
    rationale: str


class FeedRefreshMetaOut(BaseModel):
    source: str
    item_count: int
    live_item_count: int
    seeded_item_count: int
    refreshed_at: datetime
    is_stale: bool
    refresh_window_hours: int
    auto_refresh_enabled: bool


class ProgressSnapshotOut(PortalBase):
    id: int
    user_id: int
    date: date
    learning_completion_pct: float
    python_practice_count: int
    projects_completed_count: int
    interview_readiness_score: int
    notes: str

File: frontend/src/lib/types/portal.ts
export type DashboardSummary = {
  user_name: string;
  headline: string;
  learning_streak: number;
  stats: { label: string; value: string; tone: string }[];
  next_lesson: { id: number; title: string; slug: string; summary: string } | null;
  recommended_exercise: { id: number; title: string; category: string; difficulty: string } | null;
  active_projects: { id: number; title: string; status: string; portfolio_score: number }[];
  recommendation_panel: { title: string; reason: string }[];
  adaptive_focus?: AdaptiveFocus | null;
};

export type DashboardToday = {
  focus: string[];
  highlights: string[];
  blockers: string[];
};

export type Lesson = {
  id: number;
  learning_path_id: number;
  title: string;
  slug: string;
  summary: string;
  content_md: string;
  prerequisites_json: string[];
  estimated_minutes: number;
  order_index: number;
  tags_json: string[];
  is_completed?: boolean;
};

export type LearningPath = {
  id: number;
  title: string;
  slug: string;
  description: string;
  level: string;
  estimated_hours: number;
  order_index: number;
  is_active: boolean;
  tags_json: string[];
  completion_pct: number;
  lessons: Lesson[];
};

export type Course = {
  id: number;
  title: string;
  slug: string;
  description: string;
  difficulty: string;
  estimated_hours: number;
  milestones_json: {
    label: string;
    status: string;
    goal?: string;
    outcome?: string;
    path_slug?: string;
    path_title?: string;
    lesson_slugs?: string[];
    exercise_slugs?: string[];
    project_slug?: string;
    deliverable?: string;
    why_it_matters?: string;
  }[];
  status: string;
  track_focus: string;
};

export type Exercise = {
  id: number;
  title: string;
  slug: string;
  category: string;
  difficulty: string;
  prompt_md: string;
  starter_code: string;
  solution_code: string;
  explanation_md: string;
  tags_json: string[];
  progression_label?: string | null;
  practice_stage?: string | null;
  hint_md?: string | null;
  review_checklist_json?: string[];
  success_criteria_json?: string[];
  related_lesson_slugs?: string[];
  related_lesson_titles?: string[];
  next_exercise_id?: number | null;
  next_exercise_slug?: string | null;
  next_exercise_title?: string | null;
};

export type ExerciseAttempt = {
  id: number;
  exercise_id: number;
  submitted_code: string;
  notes: string;
  status: string;
  score: number;
  attempted_at: string;
};

export type ExerciseDetail = {
  exercise: Exercise;
  attempts: ExerciseAttempt[];
  review_prompt?: string | null;
};

export type KnowledgeArticle = {
  id: number;
  title: string;
  slug: string;
  category: string;
  summary: string;
  content_md: string;
  source_links_json: string[];
  tags_json: string[];
  last_reviewed_at: string;
};

export type NewsItem = {
  id: number;
  source_name: string;
  title: string;
  slug: string;
  summary: string;
  source_url: string;
  category: string;
  published_at: string;
  signal_score: number;
  tags_json: string[];
  is_saved: boolean;
  why_it_matters: string;
  suggested_action: string;
  focus_area: string;
  recommended_path_slug?: string | null;
  recommended_path_title?: string | null;
  recommended_exercise_category?: string | null;
};

export type Project = {
  id: number;
  title: string;
  slug: string;
  summary: string;
  status: string;
  category: string;
  stack_json: string[];
  architecture_md: string;
  repo_url: string | null;
  demo_url: string | null;
  lessons_learned_md: string;
  portfolio_score: number;
};

export type JobPosting = {
  id: number;
  source_name: string;
  title: string;
  slug: string;
  company_name: string;
  location: string;
  employment_type: string;
  summary: string;
  source_url: string;
  description_md: string;
  published_at: string;
  tags_json: string[];
  skill_gaps_json: string[];
  fit_score: number;
  is_saved: boolean;
  fit_summary: string;
  suggested_action: string;
  primary_gap?: string | null;
  focus_area: string;
  recommended_path_slug?: string | null;
  recommended_path_title?: string | null;
  recommended_exercise_category?: string | null;
};

export type InterviewQuestion = {
  id: number;
  category: string;
  role_type: string;
  difficulty: string;
  question_text: string;
  answer_outline_md: string;
  tags_json: string[];
  practice_count: number;
  last_practiced_at?: string | null;
  average_confidence?: number | null;
};

export type InterviewRoadmap = {
  focus_areas: string[];
  weekly_plan: string[];
  rationale: string[];
};

export type PortfolioReadiness = {
  overall_score: number;
  strongest_signals: string[];
  gaps_to_close: string[];
  next_best_moves: string[];
};

export type SkillGapInsight = {
  title: string;
  urgency: string;
  evidence: string;
  action_path: string;
};

export type MasteryArea = {
  area_slug: string;
  area_title: string;
  mastery_score: number;
  status: string;
  evidence: string;
  gap: string;
  weakest_signal: string;
  next_action_path: string;
  next_action_label: string;
};

export type AdaptiveFocus = {
  title: string;
  reason: string;
  action_path: string;
  target_kind: string;
  mastery_score: number;
  area_slug: string;
  area_title: string;
  action_label: string;
};

export type InterviewPractice = {
  question_id: number;
  practice_count: number;
  last_practiced_at: string;
  average_confidence: number;
};

export type Recommendation = {
  title: string;
  reason: string;
  action_path: string;
  source_kind: string;
};

export type JobFitAnalysis = {
  job_id: number;
  fit_score: number;
  strengths: string[];
  gaps: string[];
  rationale: string;
};

export type FeedRefreshMeta = {
  source: string;
  item_count: number;
  live_item_count: number;
  seeded_item_count: number;
  refreshed_at: string;
  is_stale: boolean;
  refresh_window_hours: number;
  auto_refresh_enabled: boolean;
};

export type MemoryCard = {
  id: number;
  front_md: string;
  back_md: string;
  category: string;
  source_kind: "lesson" | "exercise" | "interview" | "knowledge";
  source_title: string;
  difficulty: "beginner" | "intermediate" | "advanced";
  tags_json: string[];
  review_count: number;
  last_reviewed_at: string | null;
  confidence: number | null;
  next_review_at: string | null;
};

export type MemoryCardReview = {
  card_id: number;
  confidence: number;
  reviewed_at: string;
};

export type ReviewSessionSummary = {
  total_cards: number;
  reviewed: number;
  avg_confidence: number;
  strongest_category: string;
  weakest_category: string;
};
